### ISSUE TITLE
Support block-comment version of eslint-disable-line

### ISSUE THREAD

**Issue by zxqfox** [eslint/eslint#8781]
**Tell us about your environment**

* **ESLint Version:** master
* **Node Version:** any
* **npm Version:** any

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```
{ "rules": { "no-alert": [1] } }
```

**What did you do? Please include the actual source code causing the issue.**

```js
/* eslint-disable-line */ alert(1);
```

**What did you expect to happen?**

It should work like in JSCS:
```js
/* jscs:ignore */ alert(1)
```

**What actually happened? Please include the actual, raw output from ESLint.**

Error reported.

**Comment by kaicataldo** [comment]
Would it make sense to also make `eslint-disable-next-line` also available in block comments to fix [this issue](https://github.com/eslint/eslint/issues/7030)?

**Comment by zxqfox** [comment]
Yeah, absolutely.

Waiting for accepting then ;-)

**Comment by zxqfox** [comment]
Oops. Am I right that this issue already has 3 ok's?

**Comment by platinumazure** [comment]
Previous discussions on this have been "resolved" with the notion that we shouldn't have two ways to do the same thing. I would add that if we support line/block comments for some of our configuration comments, then users will be confused why we don't do so for all configuration comments.

Are we sure we want/need to do this?

**Comment by ilyavolodin** [comment]
@zxqfox Need one more 👍 to get this accepted. Since you proposed it, you count as champion. We need champion + 3 +1s

**Comment by not-an-aardvark** [comment]
This is a core change, so it would need to be approved by the TSC.

**Comment by ilyavolodin** [comment]
Oh, right, sorry about that. For some reason I was thinking that this would be a rule change.

**Comment by jeffvandyke** [comment]
I'd like to just throw out there that I think that both line and block comments should be allowed for all configuration comments, in regards to @platinumazure's comment. I remember being confused why I had to have specific types for certain configuration options, and in the case of JSX, requiring line comments are very limiting and make for messy code, which is why we have this issue. That's my vote.

**Comment by mysticatea** [comment]
Does it behave in this case?

```js
/* eslint-disable-line 


*/ alert(1);
```

**Comment by zxqfox** [comment]
@mysticatea Looks like it should disable both lines or does not work at all and throw an error or warning (the second is better for me).
Definitely need to add few spec for this.

@platinumazure There are cases in wild that can't be achieved with line-form. Also JSCS supported block-form of line-disabling directive. Actually, that's my case.

**Comment by not-an-aardvark** [comment]
I'm not sure about this.

I don't think we need to support the use case described in https://github.com/eslint/eslint/pull/8779 (using a block comment rather than a line comment to make sure line numbers in the file are preserved). Automatically inserting `eslint-disable-line` comments into a file with a tool is not the intended use case for disable comments, and in my opinion being able to do so while preserving line numbers seems like an overly strict requirement. (Allowing block disable comments wouldn't fully solve this problem either. For example, it still wouldn't be possible to insert a comment to disable a line in the middle of a long multiline template literal.)

I do think it would be nice to find a solution for the JSX case (https://github.com/eslint/eslint/issues/7030). But `eslint-disable-line` directives are supposed to unambiguously refer to a single line, and requiring line comments for them is a nice way of enforcing that. I'm not sure it's a good idea to allow people to use multiline `eslint-disable-line` comments, since that seems like it would just cause confusion about what line is being referred to.

**Comment by zxqfox** [comment]
Guys guys, if we said we support JSCS — there is nothing to discuss, JSCS supported this and ESLint should support this too.
If no — then no. But then why we deprecated JSCS and closed? We should reopen it as a fork of ESLint or something like that.

The main point we can't migrate from JSCS because of this and there are no workaround.

**Comment by platinumazure** [comment]
We promised compatibility with JSCS, not uniformity. We don't have
identical rules and options and sometimes have a different contract to do
the same thing in JSCS. I don't see commenting of lines with block comments
as core JSCS or ESLint functionality so I don't think we have to provide an
identical mechanism.

I would love to see if we can solve the use case some other way.

**Comment by zxqfox** [comment]
Well, what should I do with my case?

I literally can't do the same with ESLint. It means there are no compatibility.

**Comment by platinumazure** [comment]
Again: I would love to see if we can solve this some other way.

This is feeling like an XY problem. I don't know what actual problem you're trying to solve, but this feels like a clunky way to do it. I'm hoping we can come up with something better.

**Comment by zxqfox** [comment]
Okay, sorry. I'll pray 🙏

**Comment by kaicataldo** [comment]
> I don't know what actual problem you're trying to solve, but this feels like a clunky way to do it. I'm hoping we can come up with something better.

@platinumazure Mind explaining what it is that you don't like about this solution? It's clear that you're not satisfied with this solution, but I don't know what about it you're against.

My 2c: 
The most compelling reason for me to support this is the JSX use case (which only allows block comments), and I would also argue that it would be even more confusing to only allow those style comments in JSX but not allow them elsewhere. So it seems to me like supporting this style comment everywhere makes the most sense.

The multiline comment is an interesting case. Seems like it should treat lines that contain the beginning or end line of the comment as the lines it refers to.

I agree that a line comment would probably be clearer most of the time, but this feels like a case where we can decide on behavior, document it, and trust our users rather than limit them in this regard. And for what it's worth, this seems like a pretty reasonable style for a `eslint-disable-next-line` comment:
```
/*
  eslint-disable-next-line
 */
var foo;
```

**Comment by alberto** [comment]
> The main point we can't migrate from JSCS because of this and there are no workaround.
Can you explain why the current directives don't work for you?

**Comment by kaicataldo** [comment]
> Can you explain why the current directives don't work for you?

I'm also not entirely clear on this. Yes, you'd have to move the `disable-line` comment to the end of the line and make it a line comment, but that's an implementation detail. As far as I can tell, we have feature parity (but do correct me if I'm wrong, @zxqfox!), even though the implementation itself is slightly different.

**Comment by not-an-aardvark** [comment]
It looks like the discussion here is still ongoing. Should we still discuss it at the TSC meeting, or should we remove it from the agenda until we have a concrete proposal or we're unable to reach consensus?

**Comment by zxqfox** [comment]
> Can you explain why the current directives don't work for you?

> I'm also not entirely clear on this.

I have a couple of files that should be prepared before publishing and I can't use `// eslint-ignore-line` because expression can be on the to several lines like:
```js
myfunc(`a
b
c
d`);
// →
(something && myfunc(`a
b
c
d`));
```

Also we can't use `// eslint-ignore-next-line` because it adds new line each time.

Again, it worked well with jscs and even jshint.

**Comment by not-an-aardvark** [comment]
This was discussed in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/df5bb3c39c6a87f0e11dbb1e5c0fda7f54563972/notes/2017/2017-07-06.md#support-block-comment-version-of-eslint-disable-line). The TSC decided that we should continue discussing the topic on the issue before deciding whether to accept the proposal.

**Comment by platinumazure** [comment]
@zxqfox I'm sorry, I still don't understand why you need to preprocess your code with ESLint directives.

It seems to me that a best practice is to lint your code before doing any transforms, so that you're linting the source code that your developers are actually working on.

As an example, the [QUnit project](https://github.com/qunitjs/qunit) uses ES6 modules in source code and uses Rollup to combine into a single script file as part of the release process. So we lint the ES6 modules (what developers are working on), rather than the rolled-up code. I'm fairly sure most folks who use Webpack take the same basic approach.

Why do you need to lint code post-transformation in the first place?

**Comment by zxqfox** [comment]
Bundling into one file is another thing. I'm not testing uglified file with eslint, it's another case. But it's working right and I can't add new lines or whatever.
It's a kind of middleprocessing.
I just have some processing tools over files before pushing it into repo and I dont want to rely on to modify eslint config but looks like I have no choice.



At the same time I still don't understand why we looking for pros.
It does not break anything, it fixes few troubles, somebody will suffer if we will made this real? I don't see who and how it can hurt someone but I see how It helps.

Yes, in my specific case I can find few workarounds, but it's not the good way to fix it because it's workaround.
E.g. I can modify configs with ugly hacks and disable `max-len` rule in some environment. But I MUST do this because `eslint` does not support something that `jscs` does. It even sounds bad.

What is the objective reason we still looking for something that will prove the necessity?
What thing can be a proof?

**Comment by not-an-aardvark** [comment]
> At the same time I still don't understand why we looking for pros.
It does not break anything, it fixes few troubles, somebody will suffer if we will made this real? I don't see who and how it can hurt someone but I see how It helps.

In my opinion, the problem is that `eslint-disable-line` directives only apply to a single line, so it conceptually makes the most sense to only allow line comments for them. Otherwise, we would have to make mostly-arbitrary decisions about how the comments work with a multiline block.

> Yes, in my specific case I can find few workarounds, but it's not the good way to fix it because it's workaround.
E.g. I can modify configs with ugly hacks and disable `max-len` rule in some environment. But I MUST do this because `eslint` does not support something that `jscs` does. It even sounds bad.

I think providing compatibility with JSCS means that we would provide a way to support all JSCS use cases *somehow*. It doesn't necessarily mean that the use cases will be supported through the same syntax, or through precisely the same workflow. We might optimize ESLint for different use cases, making common cases more convenient as a tradeoff for making uncommon cases less convenient.

Since what you're doing (disabling rules inline while preprocessing) is a very uncommon use case, my opinion is that we don't need to optimize for it. ESLint already supports your use case through the workarounds that you mentioned -- we're supporting the functionality, although in a slightly less convenient manner.

**Comment by kaicataldo** [comment]
> In my opinion, the problem is that `eslint-disable-line` directives only apply to a single line, so it conceptually makes the most sense to only allow line comments for them. Otherwise, we would have to make mostly-arbitrary decisions about how the comments work with a multiline block.

This is the tricky part, in my opinion. That being said, I still support this for the JSX use case.

**Comment by zxqfox** [comment]
> Otherwise, we would have to make mostly-arbitrary decisions about how the comments work with a multiline block.

It costs less than this thread. Please.

> Since what you're doing (disabling rules inline while preprocessing) is a very uncommon use case, my opinion is that we don't need to optimize for it.

The subject is not an optimization. Please.

> ESLint already supports your use case through the workarounds that you mentioned

Disabling max-len rule for some env, or writing my own `max-len` rule that will look for `eslint-disable-line` in block-comment before reporting line. It's ugly.


Still don't see the answers:
- What is the objective reason we still looking for something that will prove the necessity?
- What thing can be a proof?

**Comment by platinumazure** [comment]
@zxqfox If you're worried about line length when adding disable comments, you can include `max-len` in your disable comment:

```js
myObj["property"]();  // eslint-disable-line dot-notation, max-len
```

Would that solve your problem?

**Comment by not-an-aardvark** [comment]
> It costs less than this thread. Please.

I'm not worried about the process of making the decisions, I'm worried about the long-term consequences of the decisions. I think allowing block `eslint-disable-line` comments would result in a slightly worse design overall due to increased user confusion, and so I wouldn't want to add them unless they provided a significant benefit to a common use case. That might apply to JSX, but it doesn't apply to uncommon case of adding directive comments when preprocessing.

**Comment by zxqfox** [comment]
@platinumazure No, I shown my trouble before:
```js
localVariable === 'blabla' ? debug`something here
...` : cb();
// →
localVariable === 'blabla' ? (process.env.DEBUG && debug`something here // I can't place here anything
...`) : cb();
```

**Comment by zxqfox** [comment]
> a slightly worse design overall due to increased user confusion

Actually, I'm still confused because I don't see a real objective reason why we should deny it.
I didn't saw it 2 years ago, and I still don't see it.
Moreover, with these examples of JSX I'm more and more believe that ESlint should support it because it's more flexible and less confusing in cases where we can't use endline-comment version (like mine and more important JSX and template-string cases).

**Comment by zxqfox** [comment]
> I'm not worried about the process of making the decisions

Oh, sorry. Hope you got me right: I meant that development of good safe block-comment versions of eslint-disable-line and eslint-disable-next-line costs less than we waste in this thread.

**Comment by platinumazure** [comment]
@zxqfox Not sure I follow. I wasn't aware block comments were parsed as such in template literals while line comments are part of the string.

```js
const str = `
  /* this is a comment and not part of the template literal? */
  // but this is part of the template literal?
`;
```

If I'm right and both of those are substrings, then I don't see how the block comment version would help in your template literal example.

> Oh, sorry. Hope you got me right: I meant that development of good safe block-comment versions of eslint-disable-line and eslint-disable-next-line costs less than we waste in this thread.

I doubt anyone disagrees with the note about development time. However, just because we can develop it, doesn't mean we should (i.e., maybe there isn't a "good, safe" block-comment version because the costs might not be worth the benefits).

**Comment by zxqfox** [comment]
@platinumazure Well

> Not sure I follow. I wasn't aware block comments were parsed as such in template literals while line comments are part of the string.

Yes, you can't put it into the template literal, but you dont need it, because you can place block-comment BEFORE it.

```js
/* any directive */ const mytmpl = `...
...`;
```

And you can't do this with line comment:
```js
// any broken syntax const mytmpl = `...
...`;
```

**Comment by zxqfox** [comment]
Just found some relative:
- #6335 - resolved as PR with additional "clearer" documentation
- #7890 - one more request for single line comments
- #6416 - abandoned and closed
- #6903, #5914 - jshint-like behaviour of line comments?
- #7030 - closed by starter, but has comment about reopening with 11 likes

And that is music for me:

> because JSHint treats line and block comments differently. A line comment applies to its containing block while a block comment does not. As we continue to have people switching from JSHint to ESLint, we don't want to provide something that is looks like it might behave the same way.

Same was for JSCS.
Was.

> I doubt anyone disagrees with the note about development time. However, just because we can develop it, doesn't mean we should (i.e., maybe there isn't a "good, safe" block-comment version because the costs might not be worth the benefits).

I'm not about writing code with business logic. Just about thinking how it should work.

After this little research I'm even more sure we should support both versions for ALL directives.
People keeps creating issues because they think it's more native.
And there are no __real__ reasons to not change anything — We should move forward.

Thank you for contribution and I don't know what I can do here but please mention me if I needed.

**Comment by kaicataldo** [comment]
We should make a decision about this. I'm still :+1: for the JSX use case. It looks like we have some team members for it and some against. Is it maybe time to bring this to the TSC?

**Comment by mysticatea** [comment]
I think that we have to clarify the specification of this feature before the decision. I think this proposal still has much ambiguity.

1. **How should it do on block comments which have line break?**
   I think it should ignore those or throw an error to avoid confusing.
    ```js
    /* eslint-disable-line
     *
     */ var foo
    ```
    ```js
    /* eslint-disable-next-line
     *
     */
    var foo
    ```
2. **How should it handle block comments which are in middle of a line?**
    I think the following code is confusing, so I'd like to ignore those or throw an error. However, it has to allow those to address the cases that the feature wants to support such as JSX case. It's my concern.
    ```js
    var /* eslint-disable-line */ foo
    ```
3. **How should it handle multiple block comments which are on a line?**
    I think those are merged, but I'm OK if it ignores those or throw an error.
    ```js
    /* eslint-disable-line no-shadow */ /* eslint-disable-line no-unused-vars */ var foo
    ```
4. **How should it handle a mix of `eslint-enable` and `eslint-disable-line`**
    In the following case, I'm not sure how it should behave for errors at the `foo`. I guess it should be disabled.
    ```js
    /* eslint-disable-line */ var /* eslint-enable */ foo
    ```
    If it's `/* eslint-disable */ var /* eslint-enable */ foo`, the errors at the `foo` are reported because the `foo` is after `/* eslint-enable */`. Above sample is similar to this.
    If it's `var /* eslint-enable */ foo //eslint-disable-line`, the errors at the `foo` are not reported because the `foo` is on the line which has `//eslint-disable-line`. I think that `//eslint-disable-line` and `/* eslint-disable/enable */` are different type of comments to help avoiding confusion.
5. **How should it handle `eslint` and `eslint-env` comments which are in line comments?**
    ```js
    // eslint no-undef: error
    //        no-unused-var: error
    ```
    Currently, we support those only on block comments. If we need to handle line comments and block comments consistently, we should consider this. I think that we should keep about this behavior as is.

In line comment case, we can write only one comment on a line, so it fits as the directive to control the state of a line. We can write block comments with many variations. It can bring ambiguity and complexity to users and developers.

We can disable reports with line comments in most cases already. We can disable reports with a `/* eslint-disable/enable */` comment pair in remaining cases. So I'm not sure that this feature has more value than the ambiguity and complexity.

**Comment by not-an-aardvark** [comment]
> 1. **How should it do on block comments which have line break?**
>    I think it should ignore those or throw an error to avoid confusing.
>     ```js
>     /* eslint-disable-line
>      *
>      */ var foo
>     ```
>     ```js
>     /* eslint-disable-next-line
>      *
>      */
>     var foo
>     ```

I think it would be best to report a problem at that location, and ignore the comment. However, it might be a breaking change to start reporting a problem for those cases, since we previously always ignored them.

We could ignore them for now, then start reporting a problem for them in the next major release.

> 2. **How should it handle block comments which are in middle of a line?**
>     I think the following code is confusing, so I'd like to ignore those or throw an error. However, it has to allow those to address the cases that the feature wants to support such as JSX case. It's my concern.
>     ```js
>     var /* eslint-disable-line */ foo
>     ```

I think it's fine to just ignore disable the entire line in that case. We already do that with `// eslint-disable-line` at the end of a line, so in my opinion it's not too confusing to to also ignore the entire line if there is a block comment in the middle.

> 3. **How should it handle multiple block comments which are on a line?**
>     I think those are merged, but I'm OK if it ignores those or throw an error.
>     ```js
>     /* eslint-disable-line no-shadow */ /* eslint-disable-line no-unused-vars */ var foo
>     ```

I think this should disable both rules, by just processing the comments separately.

We already do the same thing in a similar case:

```js
// eslint-disable-next-line no-shadow
var foo // eslint-disable-line no-unused-vars
```

> 4. **How should it handle a mix of `eslint-enable` and `eslint-disable-line`**
>     In the following case, I'm not sure how it should behave for errors at the `foo`. I guess it should be disabled.
>     ```js
>     /* eslint-disable-line */ var /* eslint-enable */ foo
>     ```
>     If it's `/* eslint-disable */ var /* eslint-enable */ foo`, the errors at the `foo` are reported because the `foo` is after `/* eslint-enable */`. Above sample is similar to this.
>     If it's `var /* eslint-enable */ foo //eslint-disable-line`, the errors at the `foo` are not reported because the `foo` is on the line which has `//eslint-disable-line`. I think that `//eslint-disable-line` and `/* eslint-disable/enable */` are different type of comments to help avoiding confusion.

We already handle similar cases:

```js
// eslint-disable-next-line
foo; /* eslint-enable */ bar
```

```js
foo; /* eslint-enable */ bar // eslint-disable-line
```

With the current behavior, reporting is disabled for `foo`, but enabled for `bar`.

---

Right now, we [preprocess](https://github.com/eslint/eslint/blob/61f109343503c13de7b742669e5df44964c9e4e7/lib/util/apply-disable-directives.js#L39-L60) all `disable-line` and `disable-next-line` directives into regular `disable` and `enable` directives:

```js
// eslint-disable-next-line
foo;

// is processed the same way as ↓

/* eslint-disable */foo;
/* eslint-enable */
```

So I think we could just do the same thing here regardless of whether the `disable-line` comment is a block.

**Comment by mysticatea** [comment]
@not-an-aardvark Sounds good to me.

**Comment by not-an-aardvark** [comment]
**TSC Summary**: This issue proposes allowing `eslint-disable-line` and `eslint-disable-next-line` directives in block comments. This would be convenient for JSX, and would allow some external code processors to add disable comments more easily.

**TSC Question**: Should we accept this proposal?

**Comment by jrpool** [comment]
An `/*eslint-disable-range: <line-offset>:<col-num>-<line-offset>:<col-num>*/` directive (with `line-offset` defined as relative to the opening `/*`) would seem to render `/*eslint-enable*/`, `/*eslint-disable*/`, `/*eslint-disable-line*/`, and `/*eslint-disable-next-line*/` directives superfluous. It would also make it possible to disable warnings on specific nodes and tokens, which at least in the case of template literals would not otherwise be possible. Finally, I believe I, at least, would find it reasonably unambiguous in all contexts. Whether there is any user demand for that feature I don’t know.

**Comment by not-an-aardvark** [comment]
This issue was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/216e2779058a2090a022edb3b0aa4e39de9e4a5c/notes/2017/2017-11-09.md#support-block-comment-version-of-eslint-disable-line).

**Comment by erindepew** [comment]
I'd be happy to take this on if no one else wants it!

**Comment by platinumazure** [comment]
@eslint/eslint-tsc Could someone write a comment, or link to an existing comment, outlining the semantics agreed to in the TSC meeting? There are a few proposals here.

**Comment by kaicataldo** [comment]
@not-an-aardvark's comment [here](https://github.com/eslint/eslint/issues/8781#issuecomment-328912584) I believe covers everything we discussed during the meeting.

To simplify, rules for `eslint-disable-line` and `eslint-disable-next-line` (both line and block) comments should be:
 * They must be single line comments
 * They apply to the entire line, regardless of position

@erindepew Please let us know if anything is unclear!

**Comment by jcrben** [comment]
My use case: I ran into a bug where a function is executed dynamically by an external library (Highcharts `formatter()` for tooltips, so I can't write my code as I'd like (less transpilation). I got around it with 4 or 5 eslint-disable-next-lines but it's annoying.

I also use typescript and I enjoy `/* tslint:disable */` and `/* tslint:enable */` to ignore a range of code.

**Comment by j-f1** [comment]
@jcrben FYI you can use `/* eslint-disable */` and `/* eslint-enable */` to do the same thing with ESLint, and you can even pass a rule or list of rules to enable/disable.
